server/src/main/java/org/elasticsearch/ExceptionsHelper.java

```diff
@@ -246,33 +244,22 @@ public static boolean reThrowIfNotNull(@Nullable Throwable e) {
         return true;
     }
 
-    static final int MAX_ITERATIONS = 1024;
-
-    /**
-     * Unwrap the specified throwable looking for any suppressed errors or errors as a root cause of the specified throwable.
-     *
-     * @param cause the root throwable
-     * @return an optional error if one is found suppressed or a root cause in the tree rooted at the specified throwable
-     */
-    public static Optional<Error> maybeError(final Throwable cause, final Logger logger) {
-        // early terminate if the cause is already an error
-        if (cause instanceof Error) {
-            return Optional.of((Error) cause);
+    @SuppressWarnings("unchecked")
+    private static <T extends Throwable> Optional<T> unwrap(Throwable cause, Predicate<Throwable> predicate) {
```

Comment:
when I mentioned naming, I was thinking that this one could e.g. be called something like `unwrapCausesAndSuppressed` (ugly, maybe you can think of something better :) ), which shouldn't be an issue since it's both new and private?

Reply:
Ah that makes sense :) I couldn't think of anything better though 850d507